src/Parallafka/CommitState.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Parallafka.KafkaConsumer;
+
+namespace Parallafka
+{
+    /// <summary>
+    /// Keeps track of messages that need to be committed.
+    /// </summary>
+    /// <typeparam name="TKey"></typeparam>
+    /// <typeparam name="TValue"></typeparam>
+    internal class CommitState<TKey, TValue>
+    {
+        /// <summary>
+        /// Maps partition number to a queue of the messages received from that partition.
+        /// The front of the queue is the earliest uncommitted message. Messages are removed once committed.
+        /// It's safe to commit a handled message provided all earlier (lower-offset) messages have also been marked
+        /// as handled.
+        /// </summary>
+        private readonly Dictionary<int, ConcurrentQueue<IKafkaMessage<TKey, TValue>>> _messagesNotYetCommittedByPartition;
+        private readonly ReaderWriterLockSlim _messagesNotYetCommittedByPartitionReaderWriterLock;
+
+        private readonly SemaphoreSlim _canQueueMessage;
+        private readonly CancellationToken _stopToken;
+
+        public CommitState(int maxMessagesQueued, CancellationToken stopToken)
+        {
+            this._canQueueMessage = new(maxMessagesQueued);
+            this._stopToken = stopToken;
+            this._messagesNotYetCommittedByPartition = new();
+            this._messagesNotYetCommittedByPartitionReaderWriterLock = new();
+        }
+
+        /// <summary>
+        /// Returns an enumeration of messages that are ready to be committed to kafka.
+        /// The commit queues are emptied as much as possible during this enumeration.
+        /// </summary>
+        public IEnumerable<IKafkaMessage<TKey, TValue>> GetMessagesToCommit()
+        {
+            List<ConcurrentQueue<IKafkaMessage<TKey, TValue>>> allQueues;
+
+            this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterReadLock();
+            try
+            {
+                allQueues = this._messagesNotYetCommittedByPartition.Values.ToList();
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            foreach (var queue in allQueues)
+            {
+                IKafkaMessage<TKey, TValue> messageToCommit = GetMessageToCommit(queue);
+
+                if (messageToCommit != null)
+                {
+                    yield return messageToCommit;
+                }
+            }
+        }
+
+        private IKafkaMessage<TKey, TValue> GetMessageToCommit(ConcurrentQueue<IKafkaMessage<TKey, TValue>> messagesInPartition)
+        {
+            IKafkaMessage<TKey, TValue> messageToCommit = null;
+            while (messagesInPartition.TryPeek(out IKafkaMessage<TKey, TValue> msg) && msg.WasHandled)
```

Comment:
Do you expect this to still be called concurrently? I think there's at least once race condition where you can dequeue messages never handled.

Reply:
No concurrency. It's only the MessageCommitter _commitBlock that ever calls this method.